libs/framework/src/framework.c

```diff
@@ -2015,67 +1934,112 @@ static void fw_handleEventRequest(celix_framework_t *framework, request_t* reque
         }
         celixThreadMutex_unlock(&framework->frameworkListenersLock);
+    } else if (event->type == CELIX_REGISTER_SERVICE_EVENT) {
+        //fw_log(framework->logger, CELIX_LOG_LEVEL_TRACE, "Handling fw service registration event for service '%s' with svc id %li", event->serviceName, event->registerServiceId);
+        service_registration_t* reg = NULL;
+        celix_status_t status;
+        if (event->factory != NULL) {
+            status = celix_serviceRegistry_registerServiceFactory(framework->registry, event->bndEntry->bnd, event->serviceName, event->factory, event->properties, event->registerServiceId, &reg);
+        } else {
+            status = celix_serviceRegistry_registerService(framework->registry, event->bndEntry->bnd, event->serviceName, event->svc, event->properties, event->registerServiceId, &reg);
+        }
+        if (status != CELIX_SUCCESS) {
+            fw_log(framework->logger, CELIX_LOG_LEVEL_ERROR, "Could not register service async. svc name is %s, error is %s", event->serviceName, celix_strerror(status));
+        } else if (event->registerCallback != NULL) {
+            event->registerCallback(event->registerData, serviceRegistration_getServiceId(reg));
+        }
+    } else if (event->type == CELIX_UNREGISTER_SERVICE_EVENT) {
+        //fw_log(framework->logger, CELIX_LOG_LEVEL_TRACE, "Handling fw service unregister event for service id %li", event->unregisterServiceId);
+        celix_serviceRegistry_unregisterService(framework->registry, event->bndEntry->bnd, event->unregisterServiceId);
+    } else if (event->type == CELIX_GENERIC_EVENT) {
+        //fw_log(framework->logger, CELIX_LOG_LEVEL_TRACE, "Handling event %s", event->eventName);
+        if (event->genericProcess != NULL) {
+            event->genericProcess(event->genericProcessData);
+        }
+    }
+
+    if (event->doneCallback != NULL) {
+        event->doneCallback(event->doneData);
     }
 }
 
-static inline void fw_handleEvents(celix_framework_t* framework, celix_array_list_t* localRequests) {
+static inline celix_framework_event_t* fw_topEventFromQueue(celix_framework_t* fw) {
+    celix_framework_event_t* e = NULL;
+    celixThreadMutex_lock(&fw->dispatcher.mutex);
+    if (fw->dispatcher.eventQueueSize > 0) {
+        e = &fw->dispatcher.eventQueue[fw->dispatcher.eventQueueFirstEntry];
+    } else if (celix_arrayList_size(fw->dispatcher.dynamicEventQueue) > 0) {
+        e = celix_arrayList_get(fw->dispatcher.dynamicEventQueue, 0);
+    }
+    celixThreadMutex_unlock(&fw->dispatcher.mutex);
+    return e;
+}
+
+static inline bool fw_removeTopEventFromQueue(celix_framework_t* fw) {
+    bool dynamicallyAllocated = false;
+    celixThreadMutex_lock(&fw->dispatcher.mutex);
+    if (fw->dispatcher.eventQueueSize > 0) {
+        fw->dispatcher.eventQueueFirstEntry = (fw->dispatcher.eventQueueFirstEntry+1) % CELIX_FRAMEWORK_STATIC_EVENT_QUEUE_SIZE;
+        fw->dispatcher.eventQueueSize -= 1;
```

Comment:
The dynamic event queue will be empty before continuing on the ring buffer, but I agree it is better to increase the ring buffer size instead. 
This is a simpler approach.

Reply:
What exactly do you mean with "simpler approach"? As far as simplicity for reading/understanding the code, this is a complex approach. It took me more than 15 minutes to understand the code related to adding/retrieving events. The difficulty comes from understanding two mechanisms instead of one, but also increases cyclomatic complexity more than necessary.

Or do you mean simple in terms of implementation? As for that, using something like https://github.com/dhess/c-ringbuf would arguably be simpler, no? I think one could even argue that making a celix-specific dynamically sized ringbuffer implementation would also be easier to test/maintain than the current code. Speaking of which, are the `celix_framework_addToEventQueue`, `fw_removeTopEventFromQueue` etc functions tested?

Reply:
I meant only using a ring buffer instead of a static ring buffer and a fallback to a dynamic array is a simpler approach.

Reply:
I agree using a existing ring buffer would be preferable. 

The c-ringbuf has no license, so this is not an option. 
I was hoping for something like freebsd queue (see `man queue`), but this has no ring buffer. 
There is a buf_ring from freebsd, but this is not "header (marcro) only". 

So for this PR I would like to keep the current ring buffer usage.

Reply:
the event queue is tested indirectly through service registration, tracking, etc.

Flooding the event queue so that the increasing of the event queue is tested is done at 
bundle_context_services_test.cpp / floodEventLoopTest

Also added an explicit test for firing a generic event

Reply:
I reverted the changes for only using a ring buffer, which then can increase in size. 
This is more difficult than I initial expected. 

Currently all the waitFor call (e.g. waitForAsyncRegistration) assume that if an event is not in the event queue, the event is completely finished. 
This is implemented by first getting the top event from the queue, handle the event (with no lock on the event queue needed) and then removing the event. 
If during this process the events are moved (i.e. the ring buffer size is increased), this will crash. 

Maybe this should be designed and implemented differently, but I would prefer a new PR for this.